lib/marked.js

```diff
@@ -552,6 +561,14 @@ var inline = {
   text: /^[\s\S]+?(?=[\\<!\[`*]|\b_| {2,}\n|$)/
 };
 
+if (NEW_TEXT) {
+  // TODO: If we replace ' {2,}\n' with '  \n' and address trailing whitespace,
+  //       we break the definition of GFM inline.breaks further down (affects the gfm_break test).
+  // Furthermore, we still have trouble with the email pattern substituted in: /|[...]+@/, which
+  // is vulnerable to REDOS just like /| {2,}\n/ was
+  inline.text = /[\s\S](?:[\\<!\[`*]|\b_|  \n|$)/;
```

Comment:
Looking for other thoughts here.

1. I think we can work with the definition of line breaks (` {2,}\n` vs. `  \n` with possibly-preceding spaces).
2. I'm not sure how to deal with the email problem. The full regex is this:

```[\s\S](?:[\\<!\[`*~]|https?:\/\/|ftp:\/\/|www\.|[a-zA-Z0-9.!#$%&'*+\/=?_`{\|}~-]+@|\b_|  \n|$)```

and the `/|[a-zA-Z0-9.!#$%&'*+\/=?_`{\|}~-]+@` is problematic without deliberately seeking out emails char-by-char.

We could do a linear pass for each of these options, but @UziTech likes regexes. This particular case seems pretty friendly to a linear pass IMO.

Reply:
Where is that email regex in `marked.js`? I don't see it

Reply:
[Here](https://github.com/markedjs/marked/blob/master/lib/marked.js#L577)

Reply:
I'm fine with doing more work in the lexer if it is going to speed things up.

Would the fix be to look for an `@` symbol before looking for those other chars, or checking for the rest of the regex if an `@` isn't found?

Reply:
I sketched an approach in 24d4a5e. I used my typical tactic of "mimic `RegExp.exec`" to reduce the visible impact. The email is handled [here](https://github.com/markedjs/marked/commit/24d4a5efb3653a676646750a67c1195959e27f93#diff-81ab0a5aa39b7a91951fc28d2d151496R587) amidst all the other offsetOfX routines. The rulesets are loaded with functions like [this](https://github.com/markedjs/marked/commit/24d4a5efb3653a676646750a67c1195959e27f93#diff-81ab0a5aa39b7a91951fc28d2d151496R656) and [this](https://github.com/markedjs/marked/commit/24d4a5efb3653a676646750a67c1195959e27f93#diff-81ab0a5aa39b7a91951fc28d2d151496R662).

Before I debug it all the way and deal with TODOs like [this](https://github.com/markedjs/marked/commit/24d4a5efb3653a676646750a67c1195959e27f93#diff-81ab0a5aa39b7a91951fc28d2d151496R733), I'd like to hear opinions about this approach strategy.

As a proof of concept, the PoC from the Sonatype folks is fine after 24d4a5e.

```
$ ./poc-sonatype.js  > /dev/null
String Length = 100005 | Execution time (seconds) 0.013
String Length = 1000005 | Execution time (seconds) 0.075
String Length = 10000005 | Execution time (seconds) 0.764
```

It shows linear growth as string becomes very large, to be expected with any O(n) algorithm. Without 24d4a53 it takes more than 10 seconds to evaluate just the first case.